src/App.tsx

```diff
@@ -1,26 +1,197 @@
+  const [errorMessage, setErrorMessage] = useState<Errors | null>(null);
+  const [selectedOption, setSelectedOption] = useState<Options>(Options.All);
+  const [focus, setFocus] = useState(false);
+
+  const filterTodos = (posts: Todo[], option: Options) => {
+    switch (option) {
+      case Options.Active:
+        return posts.filter(post => !post.completed);
+
+      case Options.Completed:
+        return posts.filter(post => post.completed);
+
+      default:
+        return posts;
+    }
+  };
+
+  const todoFilterSelectedOptions = filterTodos(todos, selectedOption);
+  const todoFilterOptionActive = filterTodos(todos, Options.Active);
+
+  const clearError = useCallback(() => {
+    setErrorMessage(null);
+  }, []);
+
+  const showError = useCallback(
+    (error: Errors) => {
+      setErrorMessage(error);
+
+      setTimeout(() => {
+        clearError();
+      }, 3000);
+    },
+    [clearError],
+  );
+
+  useEffect(() => {
+    setFocus(true);
+
+    todoService
+      .getTodos()
+      .then(setTodos)
+      .catch(() => showError(Errors.LoadTodos));
+  }, [showError]);
+
+  const addTodo = useCallback(
+    async ({ userId, title, completed }: Omit<Todo, 'id'>) => {
+      const newTempTodo: Todo = {
+        id: 0,
+        userId,
+        title,
+        completed,
+      };
+
+      setTempTodo(newTempTodo);
+
+      try {
+        const newTodo = await todoService.createTodos({
+          userId,
+          title,
+          completed,
+        });
+
+        setTodos(currentTodo => [...currentTodo, newTodo]);
+      } catch {
+        showError(Errors.AddTodo);
+        throw new Error();
+      } finally {
+        setTempTodo(null);
+        setFocus(true);
+      }
+    },
+    [showError],
+  );
+
+  const deleteTodo = useCallback(
+    async (todoId: number) => {
+      try {
+        await todoService.deleteTodos(todoId);
+        setTodos(currentTodo => currentTodo.filter(todo => todo.id !== todoId));
+        setFocus(true);
+      } catch {
+        showError(Errors.DeleteTodo);
+        throw new Error();
+      }
+    },
+    [showError],
+  );
+
+  const onToogleAll = (completed: boolean) => {
+    const updatedTodos = todos.map(todo => ({
+      ...todo,
+      completed,
+    }));
+
+    setTodos(updatedTodos);
+  };
+
+  const clearCompleted = useCallback(() => {
+    const completedTodos = todos.filter(todo => todo.completed);
+
+    completedTodos.map(todo =>
+      deleteTodo(todo.id)
+        .then(() => {
+          setTodos(currentTodos => currentTodos.filter(t => t.id !== todo.id));
+        })
+        .catch(() => {
+          showError(Errors.DeleteTodo);
+        }),
+    );
+  }, [todos, showError, deleteTodo]);
+
+  const toggleCompleted = useCallback((todoId: number) => {
+    setTodos(currentTodo =>
+      currentTodo.map(todo =>
+        todo.id === todoId ? { ...todo, completed: !todo.completed } : todo,
+      ),
+    );
+  }, []);
+
+  // function updateTodo(updatedTodo: Todo) {
```

Comment:
remove comments

It's nice you already have the solution for the next stage, but add it in the next task then or uncomment here)